diff -ur hackage-package-parsec-3.1.13.0/parsec.cabal ghc-package-parsec-3.1.13.0/parsec.cabal
--- hackage-package-parsec-3.1.13.0/parsec.cabal	2018-02-06 17:13:02.000000000 +0800
+++ ghc-package-parsec-3.1.13.0/parsec.cabal	2001-09-09 09:46:40.000000000 +0800
@@ -21,18 +21,18 @@
 license-file:   LICENSE
 author:         Daan Leijen <daan@microsoft.com>, Paolo Martini <paolo@nemail.it>, Antoine Latter <aslatter@gmail.com>
 maintainer:     Herbert Valerio Riedel <hvr@gnu.org>
-homepage:       https://github.com/hvr/parsec
-bug-reports:    https://github.com/hvr/parsec/issues
+homepage:       https://github.com/haskell/parsec
+bug-reports:    https://github.com/haskell/parsec/issues
 category:       Parsing
 
 build-type:     Simple
-tested-with:    GHC ==8.4.1 || ==8.2.2 || ==8.0.2 || ==7.10.3 || ==7.8.4 || ==7.6.3 || ==7.4.2
+tested-with:    GHC ==8.6.1 || ==8.4.3 || ==8.2.2 || ==8.0.2 || ==7.10.3 || ==7.8.4 || ==7.6.3 || ==7.4.2 || ==7.4.1
 
 extra-source-files: ChangeLog.md, README.md
 
 source-repository head
     type: git
-    location: https://github.com/hvr/parsec
+    location: https://github.com/haskell/parsec
 
 library
     hs-source-dirs: src
@@ -64,7 +64,7 @@
         Text.ParserCombinators.Parsec.Token
 
     build-depends:
-        base       >= 4.5.1   && < 4.12,
+        base       >= 4.5.0   && < 4.13,
         mtl        >= 1.1.1   && < 2.3,
         bytestring >= 0.9.2.1 && < 0.11,
         text       >= 0.11.3  && < 1.3
