diff --git a/src/Data/Singletons/Util.hs b/src/Data/Singletons/Util.hs
index 0f8f788..d5a06ef 100644
--- a/src/Data/Singletons/Util.hs
+++ b/src/Data/Singletons/Util.hs
@@ -92,7 +92,7 @@ qReportError :: Quasi q => String -> q ()
 qReportError = qReport True
 
 -- | Generate a new Unique
-qNewUnique :: DsMonad q => q Int
+qNewUnique :: DsMonad q => q Uniq
 qNewUnique = do
   Name _ flav <- qNewName "x"
   case flav of
@@ -203,7 +203,7 @@ suffixName ident symb n =
 -- convert a number into both alphanumeric and symoblic forms
 uniquePrefixes :: String   -- alphanumeric prefix
                -> String   -- symbolic prefix
-               -> Int
+               -> Uniq
                -> (String, String)  -- (alphanum, symbolic)
 uniquePrefixes alpha symb n = (alpha ++ n_str, symb ++ convert n_str)
   where
