diff --git a/src/Data/Singletons/Util.hs b/src/Data/Singletons/Util.hs
index 0f8f788..0ada2fe 100644
--- a/src/Data/Singletons/Util.hs
+++ b/src/Data/Singletons/Util.hs
@@ -96,7 +96,7 @@ qNewUnique :: DsMonad q => q Int
 qNewUnique = do
   Name _ flav <- qNewName "x"
   case flav of
-    NameU n -> return n
+    NameU n -> return (fromInteger n)
     _       -> error "Internal error: `qNewName` didn't return a NameU"
 
 checkForRep :: Quasi q => [Name] -> q ()
