diff --git i/crypton-x509-system.cabal w/crypton-x509-system.cabal
index 2085a6e..ceb7689 100644
--- i/crypton-x509-system.cabal
+++ w/crypton-x509-system.cabal
@@ -45,7 +45,7 @@ library
     if os(windows)
         exposed-modules: System.X509.Win32
         cpp-options:     -DWINDOWS
-        extra-libraries: Crypt32
+        extra-libraries: crypt32
         build-depends:
             Win32,
             crypton-asn1-encoding
