diff -ur hackage-package-deepseq-1.4.3.0/deepseq.cabal ghc-package-deepseq-1.4.3.0/deepseq.cabal
--- hackage-package-deepseq-1.4.3.0/deepseq.cabal	2017-07-20 06:00:58.000000000 +0800
+++ ghc-package-deepseq-1.4.3.0/deepseq.cabal	2001-09-09 09:46:40.000000000 +0800
@@ -61,7 +61,7 @@
   if impl(ghc>=7.8)
     other-extensions: EmptyCase
 
-  build-depends: base       >= 4.5 && < 4.11,
+  build-depends: base       >= 4.5 && < 4.12,
                  array      >= 0.4 && < 0.6
   ghc-options: -Wall
 
