diff -ur hackage-package-parsec-3.1.13.0/parsec.cabal ghc-package-parsec-3.1.13.0/parsec.cabal
--- hackage-package-parsec-3.1.13.0/parsec.cabal	2018-02-06 17:13:02.000000000 +0800
+++ ghc-package-parsec-3.1.13.0/parsec.cabal	2001-09-09 09:46:40.000000000 +0800
@@ -21,8 +21,8 @@
 license-file:   LICENSE
 author:         Daan Leijen <daan@microsoft.com>, Paolo Martini <paolo@nemail.it>, Antoine Latter <aslatter@gmail.com>
 maintainer:     Herbert Valerio Riedel <hvr@gnu.org>
-homepage:       https://github.com/hvr/parsec
-bug-reports:    https://github.com/hvr/parsec/issues
+homepage:       https://github.com/haskell/parsec
+bug-reports:    https://github.com/haskell/parsec/issues
 category:       Parsing
 
 build-type:     Simple
@@ -32,7 +32,7 @@
 
 source-repository head
     type: git
-    location: https://github.com/hvr/parsec
+    location: https://github.com/haskell/parsec
 
 library
     hs-source-dirs: src
