diff --git a/parsec.cabal b/parsec.cabal
index 9531768..7c2a399 100644
--- a/parsec.cabal
+++ b/parsec.cabal
@@ -21,8 +21,8 @@ license:        BSD3
 license-file:   LICENSE
 author:         Daan Leijen <daan@microsoft.com>, Paolo Martini <paolo@nemail.it>, Antoine Latter <aslatter@gmail.com>
 maintainer:     Herbert Valerio Riedel <hvr@gnu.org>
-homepage:       https://github.com/hvr/parsec
-bug-reports:    https://github.com/hvr/parsec/issues
+homepage:       https://github.com/haskell/parsec
+bug-reports:    https://github.com/haskell/parsec/issues
 category:       Parsing
 
 build-type:     Simple
@@ -32,7 +32,7 @@ extra-source-files: ChangeLog.md, README.md
 
 source-repository head
     type: git
-    location: https://github.com/hvr/parsec
+    location: https://github.com/haskell/parsec
 
 library
     hs-source-dirs: src
