diff --git a/aeson.cabal b/aeson.cabal
index a0ebd59..552f160 100755
--- a/aeson.cabal
+++ b/aeson.cabal
@@ -150,7 +150,7 @@ library
     dlist                >= 0.8.0.4  && < 0.9,
     hashable             >= 1.2.7.0  && < 1.4,
     scientific           >= 0.3.6.2  && < 0.4,
-    th-abstraction       >= 0.2.8.0  && < 0.4,
+    th-abstraction       >= 0.2.8.0  && < 0.5,
     uuid-types           >= 1.0.3    && < 1.1,
     vector               >= 0.12.0.1 && < 0.13
 
