From ed243586aa6884119dbfd00ac86ea87bed65e3b6 Mon Sep 17 00:00:00 2001
From: Moritz Angermann <moritz.angermann@gmail.com>
Date: Thu, 20 Jun 2024 02:57:55 +0000
Subject: [PATCH 08/12] Needs linker_verbose flag

---
 includes/rts/Flags.h | 1 +
 rts/RtsFlags.c       | 5 +++++
 2 files changed, 6 insertions(+)

diff --git a/includes/rts/Flags.h b/includes/rts/Flags.h
index 76a3c51..d389dfa 100644
--- a/includes/rts/Flags.h
+++ b/includes/rts/Flags.h
@@ -104,6 +104,7 @@ typedef struct _DEBUG_FLAGS {
     bool stable;         /* 't' */
     bool prof;           /* 'p' */
     bool linker;         /* 'l'   the object linker */
+    bool linker_verbose; /* 'L'   the object linker, output which scales with O(# symbols) */
     bool apply;          /* 'a' */
     bool stm;            /* 'm' */
     bool squeeze;        /* 'z'  stack squeezing & lazy blackholing */
diff --git a/rts/RtsFlags.c b/rts/RtsFlags.c
index e56640f..e7be439 100644
--- a/rts/RtsFlags.c
+++ b/rts/RtsFlags.c
@@ -190,6 +190,7 @@ void initRtsFlagsDefaults(void)
     RtsFlags.DebugFlags.prof            = false;
     RtsFlags.DebugFlags.apply           = false;
     RtsFlags.DebugFlags.linker          = false;
+    RtsFlags.DebugFlags.linker_verbose  = false;
     RtsFlags.DebugFlags.squeeze         = false;
     RtsFlags.DebugFlags.hpc             = false;
     RtsFlags.DebugFlags.sparks          = false;
@@ -1962,6 +1963,10 @@ static void read_debug_flags(const char* arg)
         case 'l':
             RtsFlags.DebugFlags.linker = true;
             break;
+        case 'L':
+            RtsFlags.DebugFlags.linker_verbose = true;
+            RtsFlags.DebugFlags.linker = true;
+            break;
         case 'a':
             RtsFlags.DebugFlags.apply = true;
             break;
-- 
2.33.0

