From c1b689f73e1085ced75e95593605e31d4d5a0ce4 Mon Sep 17 00:00:00 2001
From: Moritz Angermann <moritz.angermann@gmail.com>
Date: Thu, 20 Jun 2024 03:12:32 +0000
Subject: [PATCH 10/12] One more debug flag (-L)

---
 includes/dist-derivedconstants/header/DerivedConstants.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/includes/dist-derivedconstants/header/DerivedConstants.h b/includes/dist-derivedconstants/header/DerivedConstants.h
index 7df96e6..8e8ce99 100644
--- a/includes/dist-derivedconstants/header/DerivedConstants.h
+++ b/includes/dist-derivedconstants/header/DerivedConstants.h
@@ -494,7 +494,7 @@
 #define OFFSET_RtsFlags_ProfFlags_showCCSOnException 285
 #define REP_RtsFlags_ProfFlags_showCCSOnException b8
 #define RtsFlags_ProfFlags_showCCSOnException(__ptr__) REP_RtsFlags_ProfFlags_showCCSOnException[__ptr__+OFFSET_RtsFlags_ProfFlags_showCCSOnException]
-#define OFFSET_RtsFlags_DebugFlags_apply 228
+#define OFFSET_RtsFlags_DebugFlags_apply 229
 #define REP_RtsFlags_DebugFlags_apply b8
 #define RtsFlags_DebugFlags_apply(__ptr__) REP_RtsFlags_DebugFlags_apply[__ptr__+OFFSET_RtsFlags_DebugFlags_apply]
 #define OFFSET_RtsFlags_DebugFlags_sanity 223
-- 
2.33.0

