From 18e6f2630baebd71ef869e06efbf9d8d24291399 Mon Sep 17 00:00:00 2001
From: Moritz Angermann <moritz.angermann@gmail.com>
Date: Thu, 20 Jun 2024 03:28:52 +0000
Subject: [PATCH 12/12] fixup HAVE_DECL_ENVIRON

---
 libraries/base/include/HsBase.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/libraries/base/include/HsBase.h b/libraries/base/include/HsBase.h
index d588447..243d969 100644
--- a/libraries/base/include/HsBase.h
+++ b/libraries/base/include/HsBase.h
@@ -552,9 +552,9 @@ INLINE int __hscore_open(char *file, int how, mode_t mode) {
 #include <crt_externs.h>
 INLINE char **__hscore_environ(void) { return *(_NSGetEnviron()); }
 #else
-/* ToDo: write a feature test that doesn't assume 'environ' to
- *    be in scope at link-time. */
+#if !HAVE_DECL_ENVIRON
 extern char** environ;
+#endif
 INLINE char **__hscore_environ(void) { return environ; }
 #endif
 
-- 
2.33.0

