From e0673a4e134c58044316a07d7d4a21b058eddc7f Mon Sep 17 00:00:00 2001
From: Moritz Angermann <moritz.angermann@gmail.com>
Date: Thu, 20 Jun 2024 05:22:15 +0000
Subject: [PATCH] add STG_NORETURN to Stg.h

---
 includes/Stg.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/includes/Stg.h b/includes/Stg.h
index 46f71c0..bc5cc75 100644
--- a/includes/Stg.h
+++ b/includes/Stg.h
@@ -215,6 +215,8 @@
 
 #define STG_UNUSED    GNUC3_ATTRIBUTE(__unused__)
 
+#define STG_NORETURN GNU_ATTRIBUTE(__noreturn__)
+
 /* Prevent functions from being optimized.
    See Note [Windows Stack allocations] */
 #if defined(__clang__)
-- 
2.33.0

