diff --git a/libraries/ghc-heap/cbits/Stack.c b/libraries/ghc-heap/cbits/StackC.c
similarity index 100%
rename from libraries/ghc-heap/cbits/Stack.c
rename to libraries/ghc-heap/cbits/StackC.c
diff --git a/libraries/ghc-heap/ghc-heap.cabal.in b/libraries/ghc-heap/ghc-heap.cabal.in
index 2135996cd9..6b8568391d 100644
--- a/libraries/ghc-heap/ghc-heap.cabal.in
+++ b/libraries/ghc-heap/ghc-heap.cabal.in
@@ -31,7 +31,7 @@ library
   if !os(ghcjs)
     cmm-sources:      cbits/HeapPrim.cmm
                       cbits/Stack.cmm
-  c-sources:        cbits/Stack.c
+  c-sources:        cbits/StackC.c
 
   default-extensions: NoImplicitPrelude
 
