diff --git a/rts/linker/Elf.c b/rts/linker/Elf.c
index 9f623a6..d5fc91c 100644
--- a/rts/linker/Elf.c
+++ b/rts/linker/Elf.c
@@ -1056,6 +1056,11 @@ ocGetNames_ELF ( ObjectCode* oc )
                           )
                        ) {
                    /* Section 0 is the undefined section, hence > and not >=. */
+                   if (!(secno > 0 && secno < shnum)) {
+                      debugBelch("while processing %s\n", OC_INFORMATIVE_FILENAME(oc));
+                      debugBelch("while processing symbol %s (%d)\n", nm, j);
+                      barf("secno %d out of range (0, %d)\n", secno, shnum);
+                   }
                    CHECK(secno > 0 && secno < shnum);
                    /*
                    if (shdr[secno].sh_type == SHT_NOBITS) {
