diff --git a/libraries/base/base.cabal b/libraries/base/base.cabal
index 4df1bae..8e20826 100644
--- a/libraries/base/base.cabal
+++ b/libraries/base/base.cabal
@@ -390,7 +390,7 @@ Library
         -- advapi32: provides advanced kernel functions
         extra-libraries:
             wsock32, user32, shell32, mingw32, kernel32, advapi32,
-            mingwex, ws2_32, shlwapi, ole32, rpcrt4, ntdll
+            ws2_32, shlwapi, ole32, rpcrt4, ntdll
         -- Minimum supported Windows version.
         -- These numbers can be found at:
         --  https://msdn.microsoft.com/en-us/library/windows/desktop/aa383745(v=vs.85).aspx
