diff --git a/hadrian/cfg/system.config.in b/hadrian/cfg/system.config.in
index 6a891b4..a2289f8 100644
--- a/hadrian/cfg/system.config.in
+++ b/hadrian/cfg/system.config.in
@@ -18,6 +18,7 @@ merge-objects  = @MergeObjsCmd@
 system-merge-objects = @LD_STAGE0@
 objdump        = @ObjdumpCmd@
 ranlib         = @REAL_RANLIB_CMD@
+strip          = @StripCmd@
 sphinx-build   = @SPHINXBUILD@
 system-ar      = @AR_STAGE0@
 system-cc      = @CC_STAGE0@
diff --git a/hadrian/src/Oracles/Setting.hs b/hadrian/src/Oracles/Setting.hs
index 462d289..535cc5f 100644
--- a/hadrian/src/Oracles/Setting.hs
+++ b/hadrian/src/Oracles/Setting.hs
@@ -69,6 +69,7 @@ data Setting = BuildArch
              | ProjectPatchLevel
              | ProjectPatchLevel1
              | ProjectPatchLevel2
+             | Strip
              | SystemGhc
              | TargetArch
              | TargetOs
@@ -171,6 +172,7 @@ setting key = lookupSystemConfig $ case key of
     ProjectPatchLevel  -> "project-patch-level"
     ProjectPatchLevel1 -> "project-patch-level1"
     ProjectPatchLevel2 -> "project-patch-level2"
+    Strip              -> "strip"
     SystemGhc          -> "system-ghc"
     TargetArch         -> "target-arch"
     TargetArmVersion   -> "target-arm-version"
diff --git a/hadrian/src/Settings/Builders/Cabal.hs b/hadrian/src/Settings/Builders/Cabal.hs
index 2e42185..032ca72 100644
--- a/hadrian/src/Settings/Builders/Cabal.hs
+++ b/hadrian/src/Settings/Builders/Cabal.hs
@@ -85,15 +85,11 @@ commonCabalArgs :: Stage -> Args
 commonCabalArgs stage = do
   verbosity <- expr getVerbosity
   pkg       <- getPackage
+  strip     <- getSetting Strip
   package_id <- expr $ pkgIdentifier pkg
   let prefix = "${pkgroot}" ++ (if windowsHost then "" else "/..")
-  mconcat [ -- Don't strip libraries when cross compiling.
-            -- TODO: We need to set @--with-strip=(stripCmdPath :: Action FilePath)@,
-            -- and if it's @:@ disable stripping as well. As it is now, I believe
-            -- we might have issues with stripping on Windows, as I can't see a
-            -- consumer of 'stripCmdPath'.
-            -- TODO: See https://github.com/snowleopard/hadrian/issues/549.
-              flag CrossCompiling ? pure [ "--disable-executable-stripping"
+  mconcat [ notStage0 ? strip /= "" ? pure [ "--with-strip=" ++ strip ]
+            , flag CrossCompiling ? pure [ "--disable-executable-stripping"
                                          , "--disable-library-stripping" ]
             -- We don't want to strip the debug RTS
             , S.package rts ? pure [ "--disable-executable-stripping"

