diff --git a/rts/RtsSymbols.c b/rts/RtsSymbols.c
index 87fa98dd4f..8234190d7d 100644
--- a/rts/RtsSymbols.c
+++ b/rts/RtsSymbols.c
@@ -929,6 +929,7 @@
       SymI_HasProto(load_load_barrier)                                  \
       SymI_HasProto(cas)                                                \
       SymI_HasProto(_assertFail)                                        \
+      SymI_HasProto(keepCAFs)                                           \
       RTS_USER_SIGNALS_SYMBOLS                                          \
       RTS_INTCHAR_SYMBOLS
 
