diff --git a/rts/linker/elf_reloc_aarch64.c b/rts/linker/elf_reloc_aarch64.c
index 5e349f2839..11ad67aad4 100644
--- a/rts/linker/elf_reloc_aarch64.c
+++ b/rts/linker/elf_reloc_aarch64.c
@@ -319,10 +319,13 @@ relocateObjectCodeAarch64(ObjectCode * oc) {
 
             Elf_Rela *rel = &relaTab->relocations[i];
 
+            if(ELF64_R_TYPE(rel->r_info) == COMPAT_R_AARCH64_NONE)
+                continue;
+
             ElfSymbol *symbol =
                     findSymbol(oc,
                                relaTab->sectionHeader->sh_link,
-                               ELF64_R_SYM((Elf64_Xword)rel->r_info));
+                               ELF64_R_SYM(rel->r_info));
 
             CHECK(0x0 != symbol);
             if(0x0 == symbol->addr)
