diff --git a/compiler/GHC/Types/Unique/Supply.hs b/compiler/GHC/Types/Unique/Supply.hs
index f3e2b4b353..7a858a515e 100644
--- a/compiler/GHC/Types/Unique/Supply.hs
+++ b/compiler/GHC/Types/Unique/Supply.hs
@@ -225,7 +225,7 @@ mkSplitUniqSupply c
         (# s4, MkSplitUniqSupply (mask .|. u) x y #)
         }}}}
 
-#if !MIN_VERSION_GLASGOW_HASKELL(9,1,0,0)
+#if !MIN_VERSION_GLASGOW_HASKELL(8,10,0,0)
 foreign import ccall unsafe "genSym" genSym :: IO Int
 #else
 genSym :: IO Int
@@ -246,8 +246,13 @@ genSym = do
     return u
 #endif
 
+#if !MIN_VERSION_GLASGOW_HASKELL(9,2,0,0)
+foreign import ccall unsafe "&ghc_unique_counter" GenSymCounter :: Ptr Word
+foreign import ccall unsafe "&ghc_unique_inc"     GenSymInc     :: Ptr Int
+#else
 foreign import ccall unsafe "&ghc_unique_counter" ghc_unique_counter :: Ptr Word
 foreign import ccall unsafe "&ghc_unique_inc"     ghc_unique_inc     :: Ptr Int
+#endif
 
 initUniqSupply :: Word -> Int -> IO ()
 initUniqSupply counter inc = do
diff --git a/compiler/cbits/genSym.c b/compiler/cbits/genSym.c
index 8a47d77f27..a356b6352c 100644
--- a/compiler/cbits/genSym.c
+++ b/compiler/cbits/genSym.c
@@ -1,9 +1,15 @@
 #include <Rts.h>
 #include <assert.h>
 #include "Unique.h"
+#include "ghcversion.h"
 
+// These global variables have been moved into the RTS.  It allows them to be
+// shared with plugins even if two different instances of the GHC library are
+// loaded at the same time (#19940)
+#if !MIN_VERSION_GLASGOW_HASKELL(9,2,0,0) || GHC_STAGE < 2
 HsInt ghc_unique_counter = 0;
 HsInt ghc_unique_inc     = 1;
+#endif
 
 #define UNIQUE_BITS (sizeof (HsInt) * 8 - UNIQUE_TAG_BITS)
 #define UNIQUE_MASK ((1ULL << UNIQUE_BITS) - 1)
diff --git a/includes/rts/Globals.h b/includes/rts/Globals.h
index 8d8df48b21..c95e405ddc 100644
--- a/includes/rts/Globals.h
+++ b/includes/rts/Globals.h
@@ -32,3 +32,7 @@ mkStoreAccessorPrototype(LibHSghcFastStringTable)
 mkStoreAccessorPrototype(LibHSghcGlobalHasPprDebug)
 mkStoreAccessorPrototype(LibHSghcGlobalHasNoDebugOutput)
 mkStoreAccessorPrototype(LibHSghcGlobalHasNoStateHack)
+#if GHC_STAGE >= 2
+extern HsInt ghc_unique_counter;
+extern HsInt ghc_unique_inc;
+#endif
diff --git a/rts/Globals.c b/rts/Globals.c
index 5569ee7910..848581e4d2 100644
--- a/rts/Globals.c
+++ b/rts/Globals.c
@@ -107,3 +107,6 @@ mkStoreAccessor(LibHSghcFastStringTable)
 mkStoreAccessor(LibHSghcGlobalHasPprDebug)
 mkStoreAccessor(LibHSghcGlobalHasNoDebugOutput)
 mkStoreAccessor(LibHSghcGlobalHasNoStateHack)
+
+HsInt ghc_unique_counter = 0;
+HsInt ghc_unique_inc     = 1;
diff --git a/rts/RtsSymbols.c b/rts/RtsSymbols.c
index 987755651d..5a6ec21152 100644
--- a/rts/RtsSymbols.c
+++ b/rts/RtsSymbols.c
@@ -679,6 +679,8 @@ extern char **environ;
       SymI_HasProto(getOrSetLibHSghcGlobalHasPprDebug)                  \
       SymI_HasProto(getOrSetLibHSghcGlobalHasNoDebugOutput)             \
       SymI_HasProto(getOrSetLibHSghcGlobalHasNoStateHack)               \
+      SymI_HasProto(ghc_unique_counter)                                 \
+      SymI_HasProto(ghc_unique_inc)                                     \
       SymI_HasProto(genericRaise)                                       \
       SymI_HasProto(getProgArgv)                                        \
       SymI_HasProto(getFullProgArgv)                                    \
