From 9cfd6b56603563cdf1b837a85d1af2de35f2d614 Mon Sep 17 00:00:00 2001
From: Moritz Angermann <moritz.angermann@gmail.com>
Date: Tue, 15 Aug 2023 06:41:58 +0000
Subject: [PATCH] Update Generate.hs

---
 hadrian/src/Rules/Generate.hs | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/hadrian/src/Rules/Generate.hs b/hadrian/src/Rules/Generate.hs
index 93885b14e41..94c3049c30b 100644
--- a/hadrian/src/Rules/Generate.hs
+++ b/hadrian/src/Rules/Generate.hs
@@ -522,9 +522,6 @@ generateSettings = do
       queryTarget (\t -> substTooldir topdir (archOS_OS $ tgtArchOs t) (f t))
       where
         substTooldir :: String -> OS -> String -> String
-        substTooldir topdir OSMinGW32 s
-          = T.unpack $
-            T.replace (T.pack $ normalise $ topdir </> "inplace" </> "mingw") (T.pack "$tooldir/mingw") (T.pack $ normalise s)
         substTooldir _ _ s = s
 
 
-- 
GitLab

