diff --git a/linters/lint-commit-msg/lint-commit-msg.cabal b/linters/lint-commit-msg/lint-commit-msg.cabal
index 7d1dbd0fcb..861ad6f03d 100644
--- a/linters/lint-commit-msg/lint-commit-msg.cabal
+++ b/linters/lint-commit-msg/lint-commit-msg.cabal
@@ -26,4 +26,4 @@ executable lint-commit-msg
     base
        >= 4.14 && < 5,
     text
-       >= 1.2  && < 2.1
+       >= 1.2  && < 2.2
diff --git a/linters/lint-submodule-refs/lint-submodule-refs.cabal b/linters/lint-submodule-refs/lint-submodule-refs.cabal
index ce4012adfc..9ff85d2731 100644
--- a/linters/lint-submodule-refs/lint-submodule-refs.cabal
+++ b/linters/lint-submodule-refs/lint-submodule-refs.cabal
@@ -15,7 +15,7 @@ executable lint-submodule-refs
     base
        >= 4.14 && < 5,
     text
-       >= 1.2  && < 2.1,
+       >= 1.2  && < 2.2,
     linters-common
 
   ghc-options:
diff --git a/linters/lint-whitespace/lint-whitespace.cabal b/linters/lint-whitespace/lint-whitespace.cabal
index 61e376d1f9..ab49690435 100644
--- a/linters/lint-whitespace/lint-whitespace.cabal
+++ b/linters/lint-whitespace/lint-whitespace.cabal
@@ -28,4 +28,4 @@ executable lint-whitespace
     base
        >= 4.14 && < 5,
     text
-       >= 1.2  && < 2.1,
+       >= 1.2  && < 2.2,
diff --git a/linters/linters-common/linters-common.cabal b/linters/linters-common/linters-common.cabal
index 02245750dd..0bdd04a761 100644
--- a/linters/linters-common/linters-common.cabal
+++ b/linters/linters-common/linters-common.cabal
@@ -16,7 +16,7 @@ library
     base
        >= 4.14 && < 5,
     text
-       >= 1.2  && < 2.1,
+       >= 1.2  && < 2.2,
     deepseq
        >= 1.1,
 
