diff --git a/rts/Hpc.c b/rts/Hpc.c
index abf8543..fd4a153 100644
--- a/rts/Hpc.c
+++ b/rts/Hpc.c
@@ -137,7 +137,7 @@ readTix(void) {
     tmpModule -> hashNo = (unsigned int)expectWord64();
     ws();
     tmpModule -> tickCount = (int)expectWord64();
-    tmpModule -> tixArr = (StgWord64 *)calloc(tmpModule->tickCount,sizeof(StgWord64));
+    tmpModule -> tixArr = (StgWord64 *)stgCallocBytes(tmpModule->tickCount,sizeof(StgWord64), "Hpc.readTix");
     ws();
     expect('[');
     ws();
diff --git a/rts/RtsUtils.c b/rts/RtsUtils.c
index b9ddb2a..c7a4a5a 100644
--- a/rts/RtsUtils.c
+++ b/rts/RtsUtils.c
@@ -106,6 +106,11 @@ stgCallocBytes (size_t count, size_t size, char *msg)
       rtsConfig.mallocFailHook((W_) count*size, msg);
       stg_exit(EXIT_INTERNAL_ERROR);
     }
+    // If we run under qemu with jemalloc, calloc is not guaranteed
+    // to zero memory.
+    // - https://giters.com/jemalloc/jemalloc/issues/1844
+    // - https://lists.nongnu.org/archive/html/qemu-devel/2020-05/msg03119.html
+    memset(space, 0, count*size);
     return space;
 }
 
diff --git a/rts/linker/elf_plt.c b/rts/linker/elf_plt.c
index 9cd42ef..70817d8 100644
--- a/rts/linker/elf_plt.c
+++ b/rts/linker/elf_plt.c
@@ -1,4 +1,5 @@
 #include "Rts.h"
+#include "RtsUtils.h"
 #include "elf_plt.h"
 
 #include <stdbool.h>
@@ -51,7 +52,7 @@ makeStub(Section * section,
           void* * addr,
           uint8_t flags) {
 
-    Stub * s = calloc(1, sizeof(Stub));
+    Stub * s = stgCallocBytes(1, sizeof(Stub), "makeStub");
     ASSERT(s != NULL);
     s->target = *addr;
     s->flags  = flags;
