diff --git a/compiler/GHC/Linker/Loader.hs b/compiler/GHC/Linker/Loader.hs
index 151e3ba..0fe4692 100644
--- a/compiler/GHC/Linker/Loader.hs
+++ b/compiler/GHC/Linker/Loader.hs
@@ -1756,7 +1756,8 @@ searchForLibUsingGcc logger dflags so dirs = do
    if (file == so)
       then return Nothing
       else do b <- doesFileExist file -- file could be a folder (see #16063)
-              return (if b then Just file else Nothing)
+              let c = any (`isExtensionOf` file) [".o", ".a", ".so", ".dylib", ".lib", ".dll"]
+              return (if b && c then Just file else Nothing)

 -- | Retrieve the list of search directory GCC and the System use to find
 --   libraries and components. See Note [Fork/Exec Windows].
