diff --git a/rts/RtsSymbols.c b/rts/RtsSymbols.c
index 10efb2a..d8ea070 100644
--- a/rts/RtsSymbols.c
+++ b/rts/RtsSymbols.c
@@ -163,7 +163,11 @@ extern char **environ;
       SymI_HasProto(__mingw_vfprintf)                    \
       /* ^^ Need to figure out why this is needed.  */   \
       SymI_HasProto(closure_sizeW_)                      \
-      /* ^^ This one needed for cardano-prelude m(  */
+      /* ^^ This one needed for cardano-prelude m(  */   \
+      SymI_NeedsProto(_tzset)                            \
+      /* ^^ This one needed for time, tzset deprecated */\
+      SymI_NeedsProto(tzset)                             \
+      /* ^^ This one needed for unix-time           */
 #else
 #define RTS_MINGW_ONLY_SYMBOLS /**/
 #endif
