diff --git a/cabal-install/src/Distribution/Client/IndexUtils.hs b/cabal-install/src/Distribution/Client/IndexUtils.hs
index 203ff3098..a45a6d79d 100644
--- a/src/Distribution/Client/IndexUtils.hs
+++ b/src/Distribution/Client/IndexUtils.hs
@@ -269,7 +269,7 @@ getSourcePackagesAtIndexState verbosity repoCtxt mb_idxState mb_activeRepos = do
             return ()
         IndexStateTime ts0 -> do
             when (isiMaxTime isi /= ts0) $
-                if ts0 > isiMaxTime isi
+                if ts0 > isiHeadTime isi
                     then warn verbosity $
                                    "Requested index-state " ++ prettyShow ts0
                                 ++ " is newer than '" ++ unRepoName rname ++ "'!"
