diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle/Monadic.hs ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle/Monadic.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle/Monadic.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle/Monadic.hs	2001-09-09 09:46:40.000000000 +0800
@@ -101,12 +101,20 @@
                         scanl, scanl1,
                         enumFromTo, enumFromThenTo )
 
-import Data.Int  ( Int8, Int16, Int32, Int64 )
-import Data.Word ( Word8, Word16, Word32, Word, Word64 )
+import Data.Int  ( Int8, Int16, Int32 )
+import Data.Word ( Word8, Word16, Word32, Word64 )
+
+#if !MIN_VERSION_base(4,8,0)
+import Data.Word ( Word )
+#endif
 
 #include "vector.h"
 #include "MachDeps.h"
 
+#if WORD_SIZE_IN_BITS > 32
+import Data.Int  ( Int64 )
+#endif
+
 data Chunk v a = Chunk Int (forall m. (PrimMonad m, Vector v a) => Mutable v (PrimState m) a -> m ())
 
 -- | Monadic streams
@@ -888,6 +896,7 @@
                         :: Monad m => Integer -> Integer -> Bundle m v Integer   #-}
 
 
+#if WORD_SIZE_IN_BITS > 32
 
 -- FIXME: the "too large" test is totally wrong
 enumFromTo_big_int :: (Integral a, Monad m) => a -> a -> Bundle m v a
@@ -906,7 +915,6 @@
     step z | z <= y    = return $ Yield z (z+1)
            | otherwise = return $ Done
 
-#if WORD_SIZE_IN_BITS > 32
 
 {-# RULES
 
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle.hs ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Bundle.hs	2001-09-09 09:46:40.000000000 +0800
@@ -111,7 +111,7 @@
 type MBundle = M.Bundle
 
 inplace :: (forall m. Monad m => S.Stream m a -> S.Stream m b)
-	-> (Size -> Size) -> Bundle v a -> Bundle v b
+        -> (Size -> Size) -> Bundle v a -> Bundle v b
 {-# INLINE_FUSED inplace #-}
 inplace f g b = b `seq` M.fromStream (f (M.elements b)) (g (M.size b))
 
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Stream/Monadic.hs ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Stream/Monadic.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Stream/Monadic.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Stream/Monadic.hs	2001-09-09 09:46:40.000000000 +0800
@@ -89,8 +89,12 @@
                         scanl, scanl1,
                         enumFromTo, enumFromThenTo )
 
-import Data.Int  ( Int8, Int16, Int32, Int64 )
+import Data.Int  ( Int8, Int16, Int32 )
+import Data.Word ( Word8, Word16, Word32, Word64 )
+
+#if !MIN_VERSION_base(4,8,0)
 import Data.Word ( Word8, Word16, Word32, Word, Word64 )
+#endif
 
 #if __GLASGOW_HASKELL__ >= 708
 import GHC.Types ( SPEC(..) )
@@ -101,6 +105,10 @@
 #include "vector.h"
 #include "MachDeps.h"
 
+#if WORD_SIZE_IN_BITS > 32
+import Data.Int  ( Int64 )
+#endif
+
 #if __GLASGOW_HASKELL__ < 708
 data SPEC = SPEC | SPEC2
 #if __GLASGOW_HASKELL__ >= 700
@@ -1403,6 +1411,8 @@
 
 
 
+#if WORD_SIZE_IN_BITS > 32
+
 -- FIXME: the "too large" test is totally wrong
 enumFromTo_big_int :: (Integral a, Monad m) => a -> a -> Stream m a
 {-# INLINE_FUSED enumFromTo_big_int #-}
@@ -1412,8 +1422,6 @@
     step z | z <= y    = return $ Yield z (z+1)
            | otherwise = return $ Done
 
-#if WORD_SIZE_IN_BITS > 32
-
 {-# RULES
 
 "enumFromTo<Int64> [Stream]"
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Util.hs ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Util.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Fusion/Util.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Fusion/Util.hs	2001-09-09 09:46:40.000000000 +0800
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 -- |
 -- Module      : Data.Vector.Fusion.Util
 -- Copyright   : (c) Roman Leshchinskiy 2009
@@ -16,7 +17,9 @@
   delay_inline, delayed_min
 ) where
 
+#if !MIN_VERSION_base(4,8,0)
 import Control.Applicative (Applicative(..))
+#endif
 
 -- | Identity monad
 newtype Id a = Id { unId :: a }
@@ -29,7 +32,7 @@
   Id f <*> Id x = Id (f x)
 
 instance Monad Id where
-  return     = Id
+  return = pure
   Id x >>= f = f x
 
 -- | Box monad
@@ -43,7 +46,7 @@
   Box f <*> Box x = Box (f x)
 
 instance Monad Box where
-  return      = Box
+  return = pure
   Box x >>= f = f x
 
 -- | Delay inlining a function until late in the game (simplifier phase 0).
@@ -55,4 +58,3 @@
 delayed_min :: Int -> Int -> Int
 {-# INLINE [0] delayed_min #-}
 delayed_min m n = min m n
-
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Generic/Mutable.hs ghc-package-vector-0.11.0.0/Data/Vector/Generic/Mutable.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Generic/Mutable.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Generic/Mutable.hs	2001-09-09 09:46:40.000000000 +0800
@@ -584,7 +584,7 @@
 new n = BOUNDS_CHECK(checkLength) "new" n
       $ unsafeNew n >>= \v -> basicInitialize v >> return v
 
--- | Create a mutable vector of the given length. The length is not checked.
+-- | Create a mutable vector of the given length. The memory is not initialized.
 unsafeNew :: (PrimMonad m, MVector v a) => Int -> m (v (PrimState m) a)
 {-# INLINE unsafeNew #-}
 unsafeNew n = UNSAFE_CHECK(checkLength) "unsafeNew" n
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Generic.hs ghc-package-vector-0.11.0.0/Data/Vector/Generic.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Generic.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Generic.hs	2001-09-09 09:46:40.000000000 +0800
@@ -1038,7 +1038,7 @@
 imapM_ f = Bundle.mapM_ (uncurry f) . Bundle.indexed . stream
 
 -- | /O(n)/ Apply the monadic action to all elements of the vector, yielding a
--- vector of results. Equvalent to @flip 'mapM'@.
+-- vector of results. Equivalent to @flip 'mapM'@.
 forM :: (Monad m, Vector v a, Vector v b) => v a -> (a -> m b) -> m (v b)
 {-# INLINE forM #-}
 forM as f = mapM f as
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Mutable.hs ghc-package-vector-0.11.0.0/Data/Vector/Mutable.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Mutable.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Mutable.hs	2001-09-09 09:46:40.000000000 +0800
@@ -268,7 +268,7 @@
 {-# INLINE new #-}
 new = G.new
 
--- | Create a mutable vector of the given length. The length is not checked.
+-- | Create a mutable vector of the given length. The memory is not initialized.
 unsafeNew :: PrimMonad m => Int -> m (MVector (PrimState m) a)
 {-# INLINE unsafeNew #-}
 unsafeNew = G.unsafeNew
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Primitive/Mutable.hs ghc-package-vector-0.11.0.0/Data/Vector/Primitive/Mutable.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Primitive/Mutable.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Primitive/Mutable.hs	2001-09-09 09:46:40.000000000 +0800
@@ -211,7 +211,7 @@
 {-# INLINE new #-}
 new = G.new
 
--- | Create a mutable vector of the given length. The length is not checked.
+-- | Create a mutable vector of the given length. The memory is not initialized.
 unsafeNew :: (PrimMonad m, Prim a) => Int -> m (MVector (PrimState m) a)
 {-# INLINE unsafeNew #-}
 unsafeNew = G.unsafeNew
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Primitive.hs ghc-package-vector-0.11.0.0/Data/Vector/Primitive.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Primitive.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Primitive.hs	2001-09-09 09:46:40.000000000 +0800
@@ -164,7 +164,9 @@
 import Data.Data     ( Data(..) )
 import Text.Read     ( Read(..), readListPrecDefault )
 
+#if !MIN_VERSION_base(4,8,0)
 import Data.Monoid   ( Monoid(..) )
+#endif
 
 #if __GLASGOW_HASKELL__ >= 708
 import qualified GHC.Exts as Exts
@@ -773,7 +775,7 @@
 mapM_ = G.mapM_
 
 -- | /O(n)/ Apply the monadic action to all elements of the vector, yielding a
--- vector of results. Equvalent to @flip 'mapM'@.
+-- vector of results. Equivalent to @flip 'mapM'@.
 forM :: (Monad m, Prim a, Prim b) => Vector a -> (a -> m b) -> m (Vector b)
 {-# INLINE forM #-}
 forM = G.forM
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Storable/Mutable.hs ghc-package-vector-0.11.0.0/Data/Vector/Storable/Mutable.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Storable/Mutable.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Storable/Mutable.hs	2001-09-09 09:46:40.000000000 +0800
@@ -1,4 +1,4 @@
-{-# LANGUAGE CPP, DeriveDataTypeable, MultiParamTypeClasses, FlexibleInstances, ScopedTypeVariables #-}
+{-# LANGUAGE CPP, DeriveDataTypeable, FlexibleInstances, MagicHash, MultiParamTypeClasses, ScopedTypeVariables #-}
 
 -- |
 -- Module      : Data.Vector.Storable.Mutable
@@ -65,8 +65,13 @@
 import Foreign.Storable
 import Foreign.ForeignPtr
 
-#if __GLASGOW_HASKELL__ >= 605
-import GHC.ForeignPtr (mallocPlainForeignPtrBytes)
+#if __GLASGOW_HASKELL__ >= 706
+import GHC.ForeignPtr (mallocPlainForeignPtrAlignedBytes)
+#elif __GLASGOW_HASKELL__ >= 700
+import Data.Primitive.ByteArray (MutableByteArray(..), newAlignedPinnedByteArray,
+                                 unsafeFreezeByteArray)
+import GHC.Prim (byteArrayContents#, unsafeCoerce#)
+import GHC.ForeignPtr
 #endif
 
 import Foreign.Ptr
@@ -201,11 +206,26 @@
 {-# INLINE mallocVector #-}
 mallocVector :: Storable a => Int -> IO (ForeignPtr a)
 mallocVector =
-#if __GLASGOW_HASKELL__ >= 605
-    doMalloc undefined
-        where
-          doMalloc :: Storable b => b -> Int -> IO (ForeignPtr b)
-          doMalloc dummy size = mallocPlainForeignPtrBytes (size * sizeOf dummy)
+#if __GLASGOW_HASKELL__ >= 706
+  doMalloc undefined
+  where
+    doMalloc :: Storable b => b -> Int -> IO (ForeignPtr b)
+    doMalloc dummy size =
+      mallocPlainForeignPtrAlignedBytes (size * sizeOf dummy) (alignment dummy)
+#elif __GLASGOW_HASKELL__ >= 700
+  doMalloc undefined
+  where
+    doMalloc :: Storable b => b -> Int -> IO (ForeignPtr b)
+    doMalloc dummy size = do
+      arr@(MutableByteArray arr#) <- newAlignedPinnedByteArray arrSize arrAlign
+      newConcForeignPtr
+        (Ptr (byteArrayContents# (unsafeCoerce# arr#)))
+        -- Keep reference to mutable byte array until whole ForeignPtr goes out
+        -- of scope.
+        (touch arr)
+      where
+        arrSize  = size * sizeOf dummy
+        arrAlign = alignment dummy
 #else
     mallocForeignPtrArray
 #endif
@@ -293,7 +313,7 @@
 {-# INLINE new #-}
 new = G.new
 
--- | Create a mutable vector of the given length. The length is not checked.
+-- | Create a mutable vector of the given length. The memory is not initialized.
 unsafeNew :: (PrimMonad m, Storable a) => Int -> m (MVector (PrimState m) a)
 {-# INLINE unsafeNew #-}
 unsafeNew = G.unsafeNew
@@ -322,14 +342,14 @@
 -- | Grow a vector by the given number of elements. The number must be
 -- positive.
 grow :: (PrimMonad m, Storable a)
-              => MVector (PrimState m) a -> Int -> m (MVector (PrimState m) a)
+     => MVector (PrimState m) a -> Int -> m (MVector (PrimState m) a)
 {-# INLINE grow #-}
 grow = G.grow
 
 -- | Grow a vector by the given number of elements. The number must be
 -- positive but this is not checked.
 unsafeGrow :: (PrimMonad m, Storable a)
-               => MVector (PrimState m) a -> Int -> m (MVector (PrimState m) a)
+           => MVector (PrimState m) a -> Int -> m (MVector (PrimState m) a)
 {-# INLINE unsafeGrow #-}
 unsafeGrow = G.unsafeGrow
 
@@ -401,7 +421,7 @@
 -- | Copy a vector. The two vectors must have the same length and may not
 -- overlap.
 copy :: (PrimMonad m, Storable a)
-                 => MVector (PrimState m) a -> MVector (PrimState m) a -> m ()
+     => MVector (PrimState m) a -> MVector (PrimState m) a -> m ()
 {-# INLINE copy #-}
 copy = G.copy
 
@@ -422,7 +442,7 @@
 -- copied to a temporary vector and then the temporary vector was copied
 -- to the target vector.
 move :: (PrimMonad m, Storable a)
-                 => MVector (PrimState m) a -> MVector (PrimState m) a -> m ()
+     => MVector (PrimState m) a -> MVector (PrimState m) a -> m ()
 {-# INLINE move #-}
 move = G.move
 
@@ -434,9 +454,9 @@
 -- copied to a temporary vector and then the temporary vector was copied
 -- to the target vector.
 unsafeMove :: (PrimMonad m, Storable a)
-                          => MVector (PrimState m) a   -- ^ target
-                          -> MVector (PrimState m) a   -- ^ source
-                          -> m ()
+           => MVector (PrimState m) a   -- ^ target
+           -> MVector (PrimState m) a   -- ^ source
+           -> m ()
 {-# INLINE unsafeMove #-}
 unsafeMove = G.unsafeMove
 
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Storable.hs ghc-package-vector-0.11.0.0/Data/Vector/Storable.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Storable.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Storable.hs	2001-09-09 09:46:40.000000000 +0800
@@ -169,7 +169,9 @@
 import Data.Data     ( Data(..) )
 import Text.Read     ( Read(..), readListPrecDefault )
 
+#if !MIN_VERSION_base(4,8,0)
 import Data.Monoid   ( Monoid(..) )
+#endif
 
 #if __GLASGOW_HASKELL__ >= 708
 import qualified GHC.Exts as Exts
@@ -783,7 +785,7 @@
 mapM_ = G.mapM_
 
 -- | /O(n)/ Apply the monadic action to all elements of the vector, yielding a
--- vector of results. Equvalent to @flip 'mapM'@.
+-- vector of results. Equivalent to @flip 'mapM'@.
 forM :: (Monad m, Storable a, Storable b) => Vector a -> (a -> m b) -> m (Vector b)
 {-# INLINE forM #-}
 forM = G.forM
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Unboxed/Base.hs ghc-package-vector-0.11.0.0/Data/Vector/Unboxed/Base.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Unboxed/Base.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Unboxed/Base.hs	2001-09-09 09:46:40.000000000 +0800
@@ -30,10 +30,14 @@
 import Control.Monad.Primitive
 import Control.Monad ( liftM )
 
-import Data.Word ( Word, Word8, Word16, Word32, Word64 )
+import Data.Word ( Word8, Word16, Word32, Word64 )
 import Data.Int  ( Int8, Int16, Int32, Int64 )
 import Data.Complex
 
+#if !MIN_VERSION_base(4,8,0)
+import Data.Word ( Word )
+#endif
+
 #if __GLASGOW_HASKELL__ >= 707
 import Data.Typeable ( Typeable )
 #else
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Unboxed/Mutable.hs ghc-package-vector-0.11.0.0/Data/Vector/Unboxed/Mutable.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Unboxed/Mutable.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Unboxed/Mutable.hs	2001-09-09 09:46:40.000000000 +0800
@@ -149,7 +149,7 @@
 {-# INLINE new #-}
 new = G.new
 
--- | Create a mutable vector of the given length. The length is not checked.
+-- | Create a mutable vector of the given length. The memory is not initialized.
 unsafeNew :: (PrimMonad m, Unbox a) => Int -> m (MVector (PrimState m) a)
 {-# INLINE unsafeNew #-}
 unsafeNew = G.unsafeNew
diff -ur hackage-package-vector-0.11.0.0/Data/Vector/Unboxed.hs ghc-package-vector-0.11.0.0/Data/Vector/Unboxed.hs
--- hackage-package-vector-0.11.0.0/Data/Vector/Unboxed.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector/Unboxed.hs	2001-09-09 09:46:40.000000000 +0800
@@ -186,7 +186,9 @@
 
 import Text.Read     ( Read(..), readListPrecDefault )
 
+#if !MIN_VERSION_base(4,8,0)
 import Data.Monoid   ( Monoid(..) )
+#endif
 
 #if __GLASGOW_HASKELL__ >= 708
 import qualified GHC.Exts as Exts (IsList(..))
@@ -820,7 +822,7 @@
 imapM_ = G.imapM_
 
 -- | /O(n)/ Apply the monadic action to all elements of the vector, yielding a
--- vector of results. Equvalent to @flip 'mapM'@.
+-- vector of results. Equivalent to @flip 'mapM'@.
 forM :: (Monad m, Unbox a, Unbox b) => Vector a -> (a -> m b) -> m (Vector b)
 {-# INLINE forM #-}
 forM = G.forM
diff -ur hackage-package-vector-0.11.0.0/Data/Vector.hs ghc-package-vector-0.11.0.0/Data/Vector.hs
--- hackage-package-vector-0.11.0.0/Data/Vector.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/Data/Vector.hs	2001-09-09 09:46:40.000000000 +0800
@@ -150,7 +150,7 @@
   -- * Conversions
 
   -- ** Lists
-  toList, fromList, fromListN,
+  toList, Data.Vector.fromList, Data.Vector.fromListN,
 
   -- ** Other vector types
   G.convert,
@@ -187,11 +187,14 @@
 import Data.Data     ( Data(..) )
 import Text.Read     ( Read(..), readListPrecDefault )
 
-import Data.Monoid   ( Monoid(..) )
 import qualified Control.Applicative as Applicative
 import qualified Data.Foldable as Foldable
 import qualified Data.Traversable as Traversable
 
+#if !MIN_VERSION_base(4,8,0)
+import Data.Monoid   ( Monoid(..) )
+#endif
+
 #if __GLASGOW_HASKELL__ >= 708
 import qualified GHC.Exts as Exts (IsList(..))
 #endif
@@ -220,8 +223,8 @@
 
 instance Exts.IsList (Vector a) where
   type Item (Vector a) = a
-  fromList = fromList
-  fromListN = fromListN
+  fromList = Data.Vector.fromList
+  fromListN = Data.Vector.fromListN
   toList = toList
 #endif
 
@@ -297,7 +300,7 @@
 
 instance Monad Vector where
   {-# INLINE return #-}
-  return = singleton
+  return = Applicative.pure
 
   {-# INLINE (>>=) #-}
   (>>=) = flip concatMap
@@ -341,7 +344,7 @@
 
 instance Traversable.Traversable Vector where
   {-# INLINE traverse #-}
-  traverse f xs = fromList Applicative.<$> Traversable.traverse f (toList xs)
+  traverse f xs = Data.Vector.fromList Applicative.<$> Traversable.traverse f (toList xs)
 
   {-# INLINE mapM #-}
   mapM = mapM
@@ -917,7 +920,7 @@
 imapM_ = G.imapM_
 
 -- | /O(n)/ Apply the monadic action to all elements of the vector, yielding a
--- vector of results. Equvalent to @flip 'mapM'@.
+-- vector of results. Equivalent to @flip 'mapM'@.
 forM :: Monad m => Vector a -> (a -> m b) -> m (Vector b)
 {-# INLINE forM #-}
 forM = G.forM
diff -ur hackage-package-vector-0.11.0.0/include/vector.h ghc-package-vector-0.11.0.0/include/vector.h
--- hackage-package-vector-0.11.0.0/include/vector.h	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/include/vector.h	2001-09-09 09:46:40.000000000 +0800
@@ -18,4 +18,3 @@
 
 #define PHASE_STREAM  Please use "PHASE_FUSED" instead
 #define INLINE_STREAM Please use "INLINE_FUSED" instead
-
diff -ur hackage-package-vector-0.11.0.0/tests/Tests/Vector.hs ghc-package-vector-0.11.0.0/tests/Tests/Vector.hs
--- hackage-package-vector-0.11.0.0/tests/Tests/Vector.hs	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/tests/Tests/Vector.hs	2001-09-09 09:46:40.000000000 +0800
@@ -643,6 +643,7 @@
         testGroup "Data.Vector.Storable.Vector (Double)" (testNumericStorableVector (undefined :: Data.Vector.Storable.Vector Double)),
 
         testGroup "Data.Vector.Unboxed.Vector ()"       (testUnitUnboxedVector (undefined :: Data.Vector.Unboxed.Vector ())),
+        testGroup "Data.Vector.Unboxed.Vector (Bool)"       (testBoolUnboxedVector (undefined :: Data.Vector.Unboxed.Vector Bool)),
         testGroup "Data.Vector.Unboxed.Vector (Int)"    (testNumericUnboxedVector (undefined :: Data.Vector.Unboxed.Vector Int)),
         testGroup "Data.Vector.Unboxed.Vector (Double)" (testNumericUnboxedVector (undefined :: Data.Vector.Unboxed.Vector Double)),
        testGroup "Data.Vector.Unboxed.Vector (Int,Bool)" (testTupleUnboxedVector (undefined :: Data.Vector.Unboxed.Vector (Int,Bool))),
diff -ur hackage-package-vector-0.11.0.0/vector.cabal ghc-package-vector-0.11.0.0/vector.cabal
--- hackage-package-vector-0.11.0.0/vector.cabal	2015-07-13 12:26:16.000000000 +0800
+++ ghc-package-vector-0.11.0.0/vector.cabal	2001-09-09 09:46:40.000000000 +0800
@@ -38,6 +38,7 @@
 Build-Type:     Simple
 
 Extra-Source-Files:
+      changelog
       README.md
       tests/LICENSE
       tests/Setup.hs
@@ -137,9 +138,9 @@
   Install-Includes:
         vector.h
 
-  Build-Depends: base >= 4.3 && < 4.9
+  Build-Depends: base >= 4.3 && < 4.10
                , primitive >= 0.5.0.1 && < 0.7
-               , ghc-prim >= 0.2 && < 0.5
+               , ghc-prim >= 0.2 && < 0.6
                , deepseq >= 1.1 && < 1.5
 
   Ghc-Options: -O2 -Wall -fno-warn-orphans
