diff -ur hackage-package-xhtml-3000.2.2/Text/XHtml/Internals.hs ghc-package-xhtml-3000.2.2/Text/XHtml/Internals.hs
--- hackage-package-xhtml-3000.2.2/Text/XHtml/Internals.hs	2017-06-18 04:51:30.000000000 +0800
+++ ghc-package-xhtml-3000.2.2/Text/XHtml/Internals.hs	2001-09-09 09:46:40.000000000 +0800
@@ -17,6 +17,7 @@
 
 import Data.Char
 import qualified Data.Semigroup as Sem
+import qualified Data.Monoid as Mon
 
 infixr 2 +++  -- combining Html
 infixr 7 <<   -- nesting Html
@@ -67,7 +68,7 @@
 instance Sem.Semigroup Html where
     (<>) = (+++)
 
-instance Sem.Monoid Html where
+instance Mon.Monoid Html where
     mempty = noHtml
     mappend = (Sem.<>)
 
